### PR TITLE
Fix vectorized grad/2 for ops that cross the devectorization boundary

### LINKED ISSUE #1729: Vectorized gradient fails for apply_vectorized ops (Cholesky, triangular_solve, cond)
Discovered while implementing the boundary-wrapper approach for #1533. Three ops fail when their inputs are vectorized:

```elixir
# cholesky
x =
  Nx.tensor([
    [[4.0, 2.0], [2.0, 5.0]],
    [[9.0, 3.0], [3.0, 5.0]],
    [[16.0, 4.0], [4.0, 8.0]]
  ])
  |> Nx.vectorize(:batch)

Nx.Defn.grad(x, fn x -> Nx.sum(Nx.LinAlg.cholesky(x)) end)
```

```elixir
# triangular_solve with a captured non-vectorized matrix
a = Nx.tensor([[1.0, 0.0], [2.0, 3.0]])
b = Nx.tensor([[4.0, 5.0], [2.0, 3.0], [1.0, 1.0]]) |> Nx.vectorize(:batch)

Nx.Defn.grad(b, fn b -> Nx.sum(Nx.LinAlg.triangular_solve(a, b)) end)
```

```elixir
# cond with vectorized input
x = Nx.tensor([[2.0, 3.0], [-5.0, -6.0], [1.0, 1.0]]) |> Nx.vectorize(:batch)
Nx.Defn.grad(x, fn x -> Nx.sum(Nx.select(Nx.greater(x, 0), x, -x)) end)
```

**Root cause:** all three are `apply_vectorized` ops — they devectorize their inputs internally and re-vectorize the result. That internal re-vectorization is incompatible with a boundary-wrapper grad that runs the recursion in fully-devectorized space. @polvalente noted on prior #1533 work that the boundary wrapper is allowed to forbid internal re-vectorization, so the fix is in the ops themselves: rewrite each to operate without internal re-vectorization, OR document the limitation as a permanent constraint of the boundary approach.

Tests for all three exist in `nx/test/nx/defn/grad_test.exs`, in https://github.com/blasphemetheus/nx/pull/8 currently skipped with comments pointing at this root cause.

### PR DESCRIPTION
**Pull request by Chapaman**
## Summary
Fixes vectorized `Nx.Defn.grad/2` for ops that internally devectorize and re-vectorize their inputs (`cholesky`, `triangular_solve`, `select`, `cond`), as reported in #1729.
The boundary-wrapper grad from #1533 runs the backward pass in devectorized space. Failures were caused by shape/name collisions at the vectorized ↔ devectorized seam (e.g. duplicate `:batch` axis names), not by missing vectorization support in the ops themselves. This PR strengthens that boundary instead of rewriting forward `apply_vectorized` wrappers.
- **`grad(:triangular_solve)`** — re-vectorize batch axes before backward math and devectorize before returning
- **`cholesky_grad`** — use explicit batch axes in dot products (same pattern as `qr_grad`)
- **`grad(:cond)`** — normalize branch gradients, clear axis-name collisions, and replace (not append) stored grads
- **`to_grad/5`** — use `revectorize` when crossing the boundary; preserve input axis names
- **`expr.ex`** — fix vectorized `cond` zero template to use `revectorize`
- **Docs/error** — clearer message when `grad/2` is called with non-tensor-aware operators like `-x` in a raw anonymous function
Adds vectorized grad tests for `cholesky`, `triangular_solve`, `select`, `cond`, and the `-x` edge case.
Closes #1729.

### CODE REVIEW

File: nx/lib/nx/defn/grad.ex [hunk 1]
```diff
@@ -1074,7 +1199,14 @@ defmodule Nx.Defn.Grad do
   defp grad(:fft, args, ans, g, _batch_count), do: grad_fft(:fft, args, ans, g)
   defp grad(:ifft, args, ans, g, _batch_count), do: grad_fft(:ifft, args, ans, g)
 
-  defp grad(:triangular_solve, [a_input, b, opts], x_input, g, _batch_count) do
+  defp grad(:triangular_solve, [a_input, b_input, opts], x_input, g, batch_count) do
+    vectorized_axes = batch_vectorized_axes(b_input, batch_count)
+
+    a = revectorize_batch_axes(a_input, vectorized_axes)
+    _b = revectorize_batch_axes(b_input, vectorized_axes)
```
Comment (polvalente) [inline nx/lib/nx/defn/grad.ex]:
```suggestion
```

File: nx/lib/nx/defn/grad.ex [hunk 2]
```diff
@@ -1504,6 +1639,28 @@ defmodule Nx.Defn.Grad do
...
+  defp revectorize_batch_axes(%T{shape: shape} = tensor, vectorized_axes) do
+    batch_count = length(vectorized_axes)
+    {batch_sizes, rest_shape} = shape |> Tuple.to_list() |> Enum.split(batch_count)
+    expected_sizes = Enum.map(vectorized_axes, &elem(&1, 1))
+
+    if batch_sizes == expected_sizes do
+      Nx.revectorize(tensor, vectorized_axes, target_shape: List.to_tuple(rest_shape))
+    else
+      tensor
+    end
```
Comment (polvalente) [inline nx/lib/nx/defn/grad.ex]: This smells like a silent failure. In which case would a tensor come in with sizes which aren't the expected ones?

In the very least I think something where `batch_sizes` are broadcastable to `expected_sizes` might occur and we're silently passing the tensor through. (e.g. {2, 1, 3} vs {2, 10, 3})

File: nx/lib/nx/binary_backend.ex [hunk 3]
```diff
@@ -527,11 +527,13 @@ defmodule Nx.BinaryBackend do
...
         <<_::bitstring-size(^left_offset_bits),
-          left_batch_item_binary::bitstring-size(^left_batch_item_bits), _::bitstring>> =
+          left_batch_item_binary::bitstring-size(^left_batch_item_bits),
+          _::bitstring>> =
           left_binary
 
         <<_::bitstring-size(^right_offset_bits),
-          right_batch_item_binary::bitstring-size(^right_batch_item_bits), _::bitstring>> =
+          right_batch_item_binary::bitstring-size(^right_batch_item_bits),
+          _::bitstring>> =
```
Comment (polvalente) [inline nx/lib/nx/binary_backend.ex]: I think this might be stale formatting wrt elixir 1.20

File: nx/lib/nx/defn/grad.ex [hunk 4]
```diff
@@ -250,25 +278,98 @@ defmodule Nx.Defn.Grad do
...
+  defp normalize_vectorized_shape(%T{vectorized_axes: []} = t), do: t
+
+  defp normalize_vectorized_shape(%T{shape: shape, names: names, vectorized_axes: vec_axes} = t) do
+    vec_count = length(vec_axes)
+
+    if tuple_size(shape) >= vec_count and
+         Enum.take(Tuple.to_list(shape), vec_count) == Enum.map(vec_axes, &elem(&1, 1)) do
+      %{
+        t
+        | shape: shape |> Tuple.to_list() |> Enum.drop(vec_count) |> List.to_tuple(),
```
Comment (polvalente) [inline nx/lib/nx/defn/grad.ex]: instead of Enum.take + Enum.drop, we could use Enum.split to avoid double traversal

File: nx/lib/nx/defn/grad.ex [hunk 5]
```diff
@@ -1522,6 +1656,54 @@ defmodule Nx.Defn.Grad do
     end
   end
 
+  defp batch_vectorized_axes(_tensor, 0), do: []
+
+  defp batch_vectorized_axes(%T{shape: shape, names: names}, batch_count) do
+    {batch_sizes, _} = shape |> Tuple.to_list() |> Enum.split(batch_count)
+    {batch_names, _} = Enum.split(names, batch_count)
```
Comment (polvalente) [inline nx/lib/nx/defn/grad.ex]: Enum.take instead of split

File: nx/lib/nx/defn/grad.ex [hunk 6]
```diff
@@ -1522,6 +1656,54 @@ defmodule Nx.Defn.Grad do
...
+
+  defp revectorize_batch_axes(%T{shape: shape} = tensor, vectorized_axes) do
+    batch_count = length(vectorized_axes)
+    expected_sizes = Enum.map(vectorized_axes, &elem(&1, 1))
+    shape_l = Tuple.to_list(shape)
+
+    # Captured non-vectorized operands (e.g. `a` when only `b` is vectorized) do not
+    # carry a batch prefix — leave them alone. If a leading prefix is present and
+    # broadcast-compatible but not equal, fail loudly instead of silently skipping.
+    if length(shape_l) < batch_count do
```
Comment (polvalente) [inline nx/lib/nx/defn/grad.ex]: tuple_size(shape) instead of lenght(shape_l) is more efficient

**Review by polvalente — APPROVED** [review]
